samples/Common/cpp/CameraHelper_cppwinrt.cpp

```diff
@@ -0,0 +1,279 @@
+#include <winrt\Windows.Foundation.Collections.h>
+#include <winrt\Windows.Foundation.h>
+#include <winrt\Windows.Media.MediaProperties.h>
+
+using namespace winrt;
+using namespace winrt::Windows::Media::Capture;
+using namespace winrt::Windows::Media::Capture::Frames;
+using namespace winrt::Windows::Media::MediaProperties;
+
+//
+// Helper method to convert a hstring to an upper case string
+//
+std::string ToUpperString(winrt::hstring value)
+{
+    auto result = winrt::to_string(value);
+    std::transform(result.begin(), result.end(), result.begin(), ::toupper);
+    return result;
+}
+
+//
+// CameraHelper factory method that regsiters a callback for when new VideoFrames become available
+//
+CameraHelper* CameraHelper::CreateCameraHelper(winrt::delegate<std::string> failureHandler, winrt::delegate<VideoFrame> newFrameArrivedHandler)
+{
+    if (failureHandler == nullptr)
+    {
+        throw(hresult_invalid_argument(L"Error: attempting to intialize camera with a null failureHandler"));
+    }
+    if (newFrameArrivedHandler == nullptr)
+    {
+        throw(hresult_invalid_argument(L"Error: attempting to intialize camera with a null FrameArrivedHandler"));
+    }
+
+    CameraHelper* instance = new CameraHelper;
+    try
+    {
+        instance->m_signalFailure.add(failureHandler);
+        instance->m_signalFrameAvailable.add(newFrameArrivedHandler);
+        instance->Initialize();
+    }
+    catch (...)
+    {
+        if (instance != nullptr)
+        {
+            delete instance;
+        }
+        throw;
+    }
+
+    return instance;
+}
+
+//
+// Initialize camera pipeline resources and register a callback for when new VideoFrames become available.
+//
+void CameraHelper::Initialize()
+{
+    // Initialize MediaCapture with default settings in video-only streaming mode.
+    // We first try to aquire exclusive sharing mode and if we fail, we then attempt again in shared mode
+    // so that multiple instances can access the camera concurrently
+    m_mediaCapture = MediaCapture();
+    auto mediaCaptureInitializationSettings = MediaCaptureInitializationSettings();
+    mediaCaptureInitializationSettings.SharingMode(m_sharingMode);
+    mediaCaptureInitializationSettings.StreamingCaptureMode(StreamingCaptureMode::Video);
+
+    // Register a callback in case MediaCapture fails. This can happen for example if another app is using the camera and we can't get ExclusiveControl
+    m_mediaCapture.Failed({ this, &CameraHelper::MediaCapture_Failed });
+
+    m_mediaCapture.InitializeAsync(mediaCaptureInitializationSettings).get();
+
+    // Get a list of available Frame source and iterate through them to find a video preview or 
+    // a video record source with color images (and not IR, depth or other types)
+    auto frameSourceIterator = m_mediaCapture.FrameSources().First();
+    const wchar_t* cameraName;
+    do
+    {
+        auto mediaStreamType = frameSourceIterator.Current().Value().Info().MediaStreamType();
```

Comment:
Will this null deref if no frame sources are available on the system?

Reply:
no, if there are no cameras connected the previous MediaCapture::InitializeAsync() will throw and catch-throw flow to the end of the app with an error message displayed in the console

---
In reply to: [290454818](https://github.com/microsoft/WindowsVisionSkillsPreview/pull/28#discussion_r290454818) [](ancestors = 290454818)